cmdstanpy/utils.py

```diff
@@ -0,0 +1,126 @@
+"""
+Utility functions
+"""
+import json
+import subprocess
+import numpy as np
+from typing import Dict, List
+
+
+def do_command(cmd:str, cwd:str=None) -> str:
+    """
+    Spawn process, print stdout/stderr to console.
+    Throws exception on non-zero returncode.
+    """
+    proc = subprocess.Popen(
+        cmd,
+        cwd=cwd,
+        stdout=subprocess.PIPE,
+        stderr=subprocess.PIPE,
+    )
+    proc.wait()
+    stdout, stderr = proc.communicate()
+    if (proc.returncode):
+        if stderr:
+            msg = 'ERROR\n {} '.format(stderr.decode('ascii').strip())
+        raise Exception(msg)
+    if stdout:
+        return stdout.decode('ascii').strip()
+    return None
+
+
+def _rdump_array(key:str, val:np.ndarray) -> str:
+    """Flatten numpy ndarray, format as Rdump variable declaration."""
+    c = 'c(' + ', '.join(map(str, val.T.flat)) + ')'
+    if (val.size, ) == val.shape:
+        return '{key} <- {c}'.format(key=key, c=c)
+    else:
+        dim = '.Dim = c{}'.format(val.shape)
+        struct = '{key} <- structure({c}, {dim})'.format(key=key, c=c, dim=dim)
+        return struct
+
+
+def jsondump(path:str, data:Dict) -> None:
+    """Dump a dict of data to a JSON file."""
+    for key, val in data.items():
+        if isinstance(val, np.ndarray) and val.size > 1:
+            data[key] = val.tolist()
+    with open(path, 'w') as fd:
+        json.dump(data, fd)
+
+
+def rdump(path:str, data:Dict) -> None:
+    """Dump a dict of data to a R dump format file."""
+    with open(path, 'w') as fd:
+        for key, val in data.items():
+            if isinstance(val, np.ndarray) and val.size > 1:
+                line = _rdump_array(key, val)
+            elif isinstance(val, list) and len(val) > 1:
+                line = _rdump_array(key, np.asarray(val))
+            else:
+                try:
+                    val = val.flat[0]
+                except AttributeError:
+                    pass
+                line = '%s <- %s' % (key, val)
+            fd.write(line)
+            fd.write('\n')
+
+
+def scan_stan_csv(filename:str) -> Dict:
+    """Capture essential config, shape from stan_csv file."""
+    dict = {}
+    draws_found = 0
+    file_is_data = False
+    with open(filename) as fp:
+        line = fp.readline().strip()
+        while len(line) > 0 and line.startswith("#"):
```

Comment:
This function is too long; I would break up the nested stuff into functions.  Even if it makes the data munging a little more complex, this makes future modifications, to track the CmdStan csv format, less bug prone.

Reply:
broke out function to scan csv file and returns dict for config and drawset headers and shape.